common/src/services/state.service.ts

```diff
@@ -1,27 +1,1385 @@
+        if (this.getAccountFromMemory({ userId: account.userId }) !== null) {
+            this.state.accounts[account.userId] = account;
+        }
+        await this.pushAccounts();
+    }
+
+    private async scaffoldNewAccountStorage(account: Account): Promise<void> {
+        const storageAccount = await this.storageService.get<Account>(account.userId, this.defaultOnDiskLocalOptions);
+        if (storageAccount != null) {
+            storageAccount.accessToken = account.accessToken;
+            storageAccount.refreshToken = account.refreshToken;
+            account = storageAccount;
+        }
+        await this.storageService.save(account.userId, account, this.defaultOnDiskLocalOptions);
+        await this.storageService.save(account.userId, account, this.defaultOnDiskMemoryOptions);
+        await this.storageService.save(account.userId, account);
+
+        if (await this.secureStorageService.get<Account>(account.userId) == null) {
+            await this.secureStorageService.save(account.userId, account);
+        }
+    }
+
+    private async pushAccounts(): Promise<void> {
+        if (this.state?.accounts == null || Object.keys(this.state.accounts).length < 1) {
+            return;
+        }
+
+        for (const i in this.state.accounts) {
+            if (this.state.accounts[i].userId === this.state.activeUserId) {
+                this.state.accounts[i].authenticationStatus = AuthenticationStatus.Active;
+            } else {
+                const vaultTimeout = await this.getVaultTimeout({
+                    storageLocation: StorageLocation.Disk,
+                    userId: this.state.accounts[i].userId,
+                });
+                const lastActive = await this.getLastActive({
+                    storageLocation: StorageLocation.Disk,
+                    userId: this.state.accounts[i].userId,
+                });
+                const diffSeconds = ((new Date()).getTime() - lastActive) / 1000;
+                this.state.accounts[i].authenticationStatus = diffSeconds < (vaultTimeout * 60) ?
+                    AuthenticationStatus.Unlocked :
+                    AuthenticationStatus.Locked;
+            }
+        }
+        this.accounts.next(this.state.accounts);
+    }
+
+    private reconcileOptions(requestedOptions: StorageOptions, defaultOptions: StorageOptions): StorageOptions {
+        if (requestedOptions == null) {
+            return defaultOptions;
+        }
+        requestedOptions.userId = requestedOptions?.userId ?? defaultOptions.userId;
+        requestedOptions.storageLocation = requestedOptions?.storageLocation ?? defaultOptions.storageLocation;
+        requestedOptions.keySuffix = requestedOptions?.keySuffix ?? defaultOptions.keySuffix;
+        requestedOptions.useSecureStorage = requestedOptions?.useSecureStorage ?? defaultOptions.useSecureStorage;
+        requestedOptions.htmlStorageLocation = requestedOptions?.htmlStorageLocation ?? defaultOptions.htmlStorageLocation;
```

Comment:
Lodash has a nice helper for this, unfortunately I don't think we use lodash, and I don't think this warrants adding another dependency.